pkg/engine/openapi.go

```diff
@@ -0,0 +1,383 @@
+package engine
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gptscript-ai/gptscript/pkg/types"
+	"github.com/tidwall/gjson"
+)
+
+var SupportedMIMETypes = []string{"application/json", "text/plain", "multipart/form-data"}
+
+type Parameter struct {
+	Name    string `json:"name"`
+	Style   string `json:"style"`
+	Explode *bool  `json:"explode"`
+}
+
+type OpenAPIInstructions struct {
+	Server           string      `json:"server"`
+	Path             string      `json:"path"`
+	Method           string      `json:"method"`
+	BodyContentMIME  string      `json:"bodyContentMIME"`
+	QueryParameters  []Parameter `json:"queryParameters"`
+	PathParameters   []Parameter `json:"pathParameters"`
+	HeaderParameters []Parameter `json:"headerParameters"`
+	CookieParameters []Parameter `json:"cookieParameters"`
+}
+
+// runOpenAPI runs a tool that was generated from an OpenAPI definition.
+// The tool itself will have instructions regarding the HTTP request that needs to be made.
+// The tools Instructions field will be in the format "#!sys.openapi '{Instructions JSON}'",
+// where {Instructions JSON} is a JSON string of type OpenAPIInstructions.
+func (e *Engine) runOpenAPI(tool types.Tool, input string) (*Return, error) {
+	envMap := map[string]string{}
+
+	for _, env := range e.Env {
+		k, v, _ := strings.Cut(env, "=")
+		envMap[k] = v
+	}
+
+	// Extract the instructions from the tool to determine server, path, method, etc.
+	var instructions OpenAPIInstructions
+	_, inst, _ := strings.Cut(tool.Instructions, types.OpenAPIPrefix+" ")
+	inst = strings.TrimPrefix(inst, "'")
+	inst = strings.TrimSuffix(inst, "'")
+	if err := json.Unmarshal([]byte(inst), &instructions); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal tool instructions: %w", err)
+	}
+
+	// Handle path parameters
+	instructions.Path = handlePathParameters(instructions.Path, instructions.PathParameters, input)
+
+	// Parse the URL
+	u, err := url.Parse(instructions.Server + instructions.Path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse server URL %s: %w", instructions.Server+instructions.Path, err)
+	}
+
+	// Set up the request
+	req, err := http.NewRequest(instructions.Method, u.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	// Check for a bearer token (only if using HTTPS)
+	if u.Scheme == "https" {
+		// For "https://example.com" the bearer token env name would be GPTSCRIPT_EXAMPLE_COM_BEARER_TOKEN
+		bearerEnv := "GPTSCRIPT_" + strings.ToUpper(strings.Replace(u.Host, ".", "_", -1)) + "_BEARER_TOKEN"
```

Comment:
I think you want u.Hostname() so that you don't get the port part if there is a port, like `:8443`

Reply:
Thanks for finding that!

Reply:
Fixed